sounddevice.py

```diff
@@ -2349,6 +2399,157 @@ def __init__(self, channel_selectors):
+
+        >>> import sounddevice as sd
+        >>> maccore_out = sd.MacCoreSettings(channel_selectors=[8, 9])
+        >>> sd.play(..., extra_settings=maccore_out)
+
+        Setting default output channels:
+
+        >>> sd.default.extra_settings = maccore_out
+        >>> sd.play(...)
+
+        Setting input channels as well:  (Input from 'Mic 9' and 'Mic 10' on RME Fireface UFX+)
+
+        >>> maccore_in = sd.MacCoreSettings(channel_selectors=[8, 9])
+        >>> sd.default.extra_settings = maccore_in, maccore_out
+        >>> sd.playrec(..., channels=1, ...)
+
+        """
+        # Check inputs:
+        conversion_dict = {
+            'Min':    _lib.paMacCoreConversionQualityMin   ,
+            'Low':    _lib.paMacCoreConversionQualityLow   ,
+            'Medium': _lib.paMacCoreConversionQualityMedium,
+            'High':   _lib.paMacCoreConversionQualityHigh  ,
+            'Max':    _lib.paMacCoreConversionQualityMax   ,
+            }
+        if rate_conv_quality not in conversion_dict:
+            raise ValueError('rate_conv_quality must be one of {!r}'.format(list(conversion_dict.keys())))
+
+        if channel_selectors is not None:
+            # Verify it is iterable:
+            if not isinstance(channel_selectors, (list, tuple)):
+                try:
+                    # Test if we can convert this to a tuple:  (or a list would work as well)
+                    channel_selectors = tuple(channel_selectors)
+                except TypeError:
+                    raise TypeError('channel_selectors must be an iterable (such as list or tuple) of int')
+            # Verify it only contains int's:
+            if not all(isinstance(ch, int) for ch in channel_selectors):
+                raise TypeError('channel_selectors must be an iterable (such as list or tuple) of int')
+
+        # Setup for our _streaminfo method to generate CFFI streaminfo struct
+        # on the fly:
+        self._cache_cffi_streaminfo = {}
+        self._cache_cffi_channelmap = {}
+
+        self._flags = conversion_dict[rate_conv_quality]
+        if change_dev_params:
+            self._flags |= _lib.paMacCoreChangeDeviceParameters
+        if not allow_rate_conv:
+            self._flags |= _lib.paMacCoreFailIfConversionRequired
+
+        self._channel_selectors = channel_selectors
+
+
+    def _streaminfo(self, *_, **kwargs):
+        # Rather than include kind, device, channels in our argument
+        # list, we'll require all inputs to be passed by keyword,
+        # through kwargs.
+        del _
+        kind = kwargs['kind']
+        device = kwargs['device']
+        channels = kwargs['channels']
+        # This is to allow some flexibility for other Host-API-specific
+        # settings such as AsioSettings or WasapiSettings and not allow
+        # any accidental reliance on positional parameters.
+
+        if not isinstance(channels, int):
```

Comment:
It seems like the _StreamBase constructor should be taking care of these kinds of checks. If you really want to be sure channels is valid, then modify the _StreamBase constructor so this function doesn't get called until after the stream object is opened.